goro/src/main/java/com/stanfy/enroscar/goro/BufferedGoroDelegate.java

```diff
@@ -0,0 +1,357 @@
+package com.stanfy.enroscar.goro;
+
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.stanfy.enroscar.goro.GoroFuture.IMMEDIATE;
+import static com.stanfy.enroscar.goro.Util.checkMainThread;
+
+/**
+ * Either delegates a call to a wrapped Goro or buffers invocation data.
+ */
+abstract class BufferedGoroDelegate extends Goro {
+
+  /** Delegate instance. */
+  private Goro delegate;
+
+  /** Temporal array of listeners that must be added after getting service connection. */
+  private final BaseListenersHandler scheduledListeners = new BaseListenersHandler(2);
+
+  /** Postponed data. */
+  private final ArrayList<Postponed> postponed = new ArrayList<>(7);
+
+  /** Protects service instance. */
+  private final Object lock = new Object();
+
+  protected boolean updateDelegate(final Goro delegate) {
```

Comment:
`delegate()` is package-visible because it must be used for tests only (to check that we actually unbound from a service).
Access to the delegate must be carefully synchronized. Hence, it's not exposed. All the delegate-related logic is handled in this class.
Classes that extend it manage service connection only.